cpp/harvest/marc_issn_lookup.cc

```diff
@@ -106,79 +115,91 @@ void UpdateSubfield(MARC::Subfields &subfields, const SubFieldInfo &sub_field_in
+}
+
+
+bool IsInISSNs(const std::vector<std::string> &issns, const std::vector<std::string> &issns_input) {
+    for (const auto &issn : issns_input) {
+        if (std::find(issns.begin(), issns.end(), StringUtil::ASCIIToUpper(issn)) != issns.end())
+            return true;
+    }
+    return false;
+}
+
+
+// avoiding duplication in issns's cache
+void InsertIfNotExist(const std::vector<std::string> &issns_input, std::vector<std::string> &issns) {
+    for (const auto &issn : issns_input) {
+        if (std::find(issns.begin(), issns.end(), StringUtil::ASCIIToUpper(issn)) != issns.end())
+            issns.emplace_back(StringUtil::ASCIIToUpper(issn));
     }
-    LOG_INFO("Online (valid): " + std::to_string(online_valid) + ", " + "Online (invalid): " + std::to_string(online_invalid) + ", "
-             + "Print (valid): " + std::to_string(print_valid) + ", " + "Print (invalid): " + std::to_string(print_invalid));
 }
 
 
-std::map<std::string, SubFieldInfo> BuildJournalCache(const std::string &input_journal_filename) {
-    std::map<std::string, SubFieldInfo> journal_cache;
+std::vector<SubFieldInfo> BuildJournalCache(const std::string &input_journal_filename) {
+    std::vector<SubFieldInfo> journal_cache;
     auto input_journal_file(MARC::Reader::Factory(input_journal_filename));
-    int record_counter(0);
+
 
     while (MARC::Record record = input_journal_file->read()) {
         SubFieldInfo sub_field_info_of_record(record);
-
-        // if sub_info is exist in the cache
-        if (journal_cache.find(sub_field_info_of_record.x_) != journal_cache.end()) {
-            if (journal_cache[sub_field_info_of_record.x_].is_online_) {
-                if (sub_field_info_of_record.is_online_)
-                    ++journal_cache[sub_field_info_of_record.x_].online_version_counter_;
-                else
-                    ++journal_cache[sub_field_info_of_record.x_].printed_version_counter_;
-            } else {
-                if (sub_field_info_of_record.is_online_) {
-                    sub_field_info_of_record.printed_version_counter_ = journal_cache[sub_field_info_of_record.x_].printed_version_counter_;
-                    journal_cache[sub_field_info_of_record.x_] = sub_field_info_of_record;
-                    ++journal_cache[sub_field_info_of_record.x_].online_version_counter_;
-                } else
-                    ++journal_cache[sub_field_info_of_record.x_].printed_version_counter_;
+        bool exist_in_journal_cache(false);
+
+        for (auto &elemt : journal_cache) {
+            if (IsInISSNs(sub_field_info_of_record.issns, elemt.issns)) {
+                if (elemt.is_online_) {
+                    if (sub_field_info_of_record.is_online_) {
+                        ++elemt.online_version_counter_;
+                        InsertIfNotExist(sub_field_info_of_record.issns, elemt.issns);
+                        elemt.is_valid_ = false;
+                        exist_in_journal_cache = true;
```

Comment:
`exist_in_journal_cache` is set to true in all cases, so we can move it 2 levels higher and only set it once.